File: components/test-support/src/test/java/org/cloudfoundry/credhub/Oauth2TestConfiguration.java
package org.cloudfoundry.credhub;

import org.springframework.boot.autoconfigure.security.oauth2.resource.ResourceServerProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.ObjectPostProcessor;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.oauth2.provider.error.DefaultWebResponseExceptionTranslator;
import org.springframework.security.oauth2.provider.error.WebResponseExceptionTranslator;
import org.springframework.security.oauth2.provider.token.DefaultAccessTokenConverter;
import org.springframework.security.oauth2.provider.token.DefaultTokenServices;
import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices;
import org.springframework.security.oauth2.provider.token.TokenStore;
import org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter;
import org.springframework.security.oauth2.provider.token.store.JwtTokenStore;

@Configuration
@Profile("unit-test")
public class Oauth2TestConfiguration {

  @Bean
  public ResourceServerProperties resourceServerProperties() {
    return new ResourceServerProperties();
  }

  @Bean
  public JwtAccessTokenConverter jwtAccessTokenConverter() throws Exception {
    final DefaultAccessTokenConverter defaultAccessTokenConverter = new DefaultAccessTokenConverter();
    defaultAccessTokenConverter.setIncludeGrantType(true);
    final JwtAccessTokenConverter jwtAccessTokenConverter = new JwtAccessTokenConverter();
    jwtAccessTokenConverter.setAccessTokenConverter(defaultAccessTokenConverter);
    jwtAccessTokenConverter.setVerifierKey(resourceServerProperties().getJwt().getKeyValue());
    jwtAccessTokenConverter.afterPropertiesSet();
    return jwtAccessTokenConverter;
  }

  @Bean
  public TokenStore tokenStore(final JwtAccessTokenConverter jwtAccessTokenConverter) {
    return new JwtTokenStore(jwtAccessTokenConverter);
  }

  @Bean
  public ResourceServerTokenServices resourceServerTokenServices(final TokenStore tokenStore) {
    final DefaultTokenServices defaultTokenServices = new DefaultTokenServices();
    defaultTokenServices.setTokenStore(tokenStore);
    return defaultTokenServices;
  }

  @Bean
  public AuthenticationManagerBuilder authenticationManagerBuilder() {
    final ObjectPostProcessor<Object> objectPostProcessor = new ObjectPostProcessor<Object>() {
      @Override
      public <O extends Object> O postProcess(final O object) {
        return object;
      }
    };
    final AuthenticationManagerBuilder authenticationManagerBuilder =
      new AuthenticationManagerBuilder(objectPostProcessor);
    authenticationManagerBuilder.parentAuthenticationManager(authenticationManager());
    return authenticationManagerBuilder;
  }

  @Bean
  public AuthenticationManager authenticationManager() {
    return authentication -> authentication;
  }

  @Bean
  public WebResponseExceptionTranslator webResponseExceptionTranslator() {
    return new DefaultWebResponseExceptionTranslator();
  }

}

File: components/test-support/src/test/kotlin/org/cloudfoundry/credhub/Oauth2TestConfiguration.kt
package org.cloudfoundry.credhub

import org.springframework.boot.autoconfigure.security.oauth2.resource.ResourceServerProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.annotation.ObjectPostProcessor
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.oauth2.provider.token.DefaultAccessTokenConverter
import org.springframework.security.oauth2.provider.token.DefaultTokenServices
import org.springframework.security.oauth2.provider.token.ResourceServerTokenServices
import org.springframework.security.oauth2.provider.token.TokenStore
import org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter
import org.springframework.security.oauth2.provider.token.store.JwtTokenStore

@Configuration
@Profile("unit-test")
class Oauth2TestConfiguration {
    @Bean
    fun resourceServerProperties(): ResourceServerProperties = ResourceServerProperties()

    @Bean
    @Throws(Exception::class)
    fun jwtAccessTokenConverter(): JwtAccessTokenConverter =
        JwtAccessTokenConverter()
            .apply {
                accessTokenConverter =
                    DefaultAccessTokenConverter()
                        .apply { setIncludeGrantType(true) }
                setVerifierKey(resourceServerProperties().jwt.keyValue)
                afterPropertiesSet()
            }

    @Bean
    fun tokenStore(jwtAccessTokenConverter: JwtAccessTokenConverter): TokenStore = JwtTokenStore(jwtAccessTokenConverter)

    @Bean
    fun resourceServerTokenServices(tokenStore: TokenStore): ResourceServerTokenServices =
        DefaultTokenServices().apply { setTokenStore(tokenStore) }

    @Bean
    fun authenticationManagerBuilder(): AuthenticationManagerBuilder {
        val objectPostProcessor =
            object : ObjectPostProcessor<Any> {
                override fun <O : Any> postProcess(dataObject: O): O = dataObject
            }
        return AuthenticationManagerBuilder(objectPostProcessor)
            .parentAuthenticationManager(authenticationManager())
    }

    @Bean
    fun authenticationManager(): AuthenticationManager = AuthenticationManager { authentication -> authentication }
}
